File: src/java/pulse/engine/PulseScenarioExec.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package pulse.engine;

import pulse.cdm.engine.SEEventHandler;
import pulse.cdm.scenario.SEScenario;
import pulse.utilities.Log;
import pulse.utilities.LogListener;

/**
 * This is a class that will run a scenario through Pulse
 * and create it's results file. You could easily write your
 * own exec class that does what you want in between actions
 * but this may work for most of your needs.
 * @author abray
 */
public class PulseScenarioExec extends Pulse
{
  public void runScenario(String logFile, SEScenario scenario, String resultsFile, CDMUpdatedCallback callback)
  {    
    if(callback != null)
      callback.drMgr = scenario.getDataRequestManager();    
    runScenario(logFile,SEScenario.unload(scenario).toString(),resultsFile,callback);
  }
  
  public void runScenario(String logFile, String scenario, String resultsFile)
  {
    runScenario(logFile, scenario, resultsFile, null);
  }
  
  public void runScenario(String logFile, String scenario, String resultsFile, CDMUpdatedCallback callback)
  {    
    double callbackFreq_s = 0;
    if(callback!=null)
    { 
      LogListener l = this.listener;
      SEEventHandler eh = this.eventManager.getHandler();
      reset();// Only create our CDM objects if we have a callback to fill data out (also save off listener and event handler)
      this.listener = l;
      this.eventManager.forwardEvents(eh);
      this.cdmCallback = callback;
      callbackFreq_s = callback.callbackFrequency_s;
      this.requestData(callback.drMgr);
    }
    this.nativeObj = nativeAllocate(logFile);
    nativeExecuteScenario(this.nativeObj, scenario, resultsFile, callbackFreq_s);
    nativeDelete(this.nativeObj);
    this.nativeObj=0;
    this.cdmCallback = null;
  }
  
  public void cancelScenario()
  {
    if(this.nativeObj!=0)
      nativeCancelScenario(this.nativeObj);
    Log.info("Cancelling execution");
  }
  
  protected native void nativeExecuteScenario(long nativeObj, String scenario, String resultsFile, double updateFreq_s);
  protected native void nativeCancelScenario(long nativeObj);
  
}

File: src/python/pulse/engine/PulseScenarioExec.py
# Distributed under the Apache License, Version 2.0.
# See accompanying NOTICE file for details.

import logging

from pulse.cdm.scenario import SEScenarioExec
from pulse.engine.PulseEngine import eModelType, eSerializationFormat


_pulse_logger = logging.getLogger('pulse')


class PulseScenarioExec(SEScenarioExec):
    __slots__ = ["_model_type"]

    def __init__(self):
        super().__init__()

    def clear(self) -> None:
        super().clear()
        self._model_type = eModelType.HumanAdultWholeBody

    def get_model_type(self) -> eModelType:
        return self._model_type
    def set_model_type(self, t: eModelType) -> None:
        self._model_type = t

    def execute_scenario(self) -> bool:
        from pulse.engine.io.PulseScenarioExec import serialize_pulse_scenario_exec_to_string
        import PyPulse

        json = serialize_pulse_scenario_exec_to_string(self, eSerializationFormat.JSON)
        return PyPulse.execute_scenario(json, PyPulse.serialization_format.json)
